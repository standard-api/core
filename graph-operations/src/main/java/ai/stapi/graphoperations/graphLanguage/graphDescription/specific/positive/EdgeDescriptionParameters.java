package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescriptionParameters;

public class EdgeDescriptionParameters implements GraphDescriptionParameters {

  private String edgeType;

  private EdgeDescriptionParameters() {
  }

  public EdgeDescriptionParameters(String edgeType) {
    this.edgeType = edgeType;
  }

  public String getEdgeType() {
    return edgeType;
  }
}
