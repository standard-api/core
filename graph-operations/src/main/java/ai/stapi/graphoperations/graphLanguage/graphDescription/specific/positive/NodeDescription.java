package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import java.util.ArrayList;
import java.util.List;

public class NodeDescription extends AbstractNodeDescription {

  public static final String SERIALIZATION_TYPE = "NodeDescription";

  protected NodeDescription() {

  }

  public NodeDescription(
      NodeDescriptionParameters parameters,
      List<GraphDescription> childDeclarations
  ) {
    super(SERIALIZATION_TYPE, childDeclarations);
    this.parameters = parameters;
  }

  public NodeDescription(NodeDescriptionParameters parameters) {
    super(SERIALIZATION_TYPE, new ArrayList<>());
    this.parameters = parameters;
  }

  public NodeDescription(
      NodeDescriptionParameters parameters,
      GraphDescription... childDeclarations
  ) {
    super(SERIALIZATION_TYPE, parameters, childDeclarations);
  }

  public NodeDescription(
      String nodeType,
      GraphDescription... childDeclarations
  ) {
    super(SERIALIZATION_TYPE, new NodeDescriptionParameters(nodeType), childDeclarations);
  }
}
