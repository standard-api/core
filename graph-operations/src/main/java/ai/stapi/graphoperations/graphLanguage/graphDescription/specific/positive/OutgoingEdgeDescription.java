package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import java.util.List;

public class OutgoingEdgeDescription extends AbstractEdgeDescription {

  public static final String SERIALIZATION_TYPE = "OutgoingEdgeDescription";

  protected OutgoingEdgeDescription() {

  }

  public OutgoingEdgeDescription(
      EdgeDescriptionParameters parameters,
      List<GraphDescription> childDeclarations
  ) {
    super(parameters, SERIALIZATION_TYPE, childDeclarations);
  }

  public OutgoingEdgeDescription(
      EdgeDescriptionParameters parameters,
      GraphDescription... childDeclarations
  ) {
    super(parameters, SERIALIZATION_TYPE, childDeclarations);
  }

  public OutgoingEdgeDescription(EdgeDescriptionParameters parameters) {
    super(parameters, SERIALIZATION_TYPE);
  }

  @Override
  public boolean isOutgoing() {
    return true;
  }
}
