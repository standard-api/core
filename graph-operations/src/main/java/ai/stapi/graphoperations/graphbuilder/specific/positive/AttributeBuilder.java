package ai.stapi.graphoperations.graphbuilder.specific.positive;

import ai.stapi.graph.attribute.Attribute;
import ai.stapi.graph.attribute.attributeFactory.AttributeValueFactoryInput;
import ai.stapi.graph.attribute.attributeFactory.GenericAttributeFactory;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class AttributeBuilder {

  private String attributeName;
  private String structureType;
  private List<AttributeValueFactoryInput> attributeValues;

  public AttributeBuilder() {
    this.attributeValues = new ArrayList<>();
  }

  public AttributeBuilder setAttributeName(String attributeName) {
    this.attributeName = attributeName;
    return this;
  }

  public AttributeBuilder addAttributeValue(AttributeValueFactoryInput attributeValue) {
    this.attributeValues.add(attributeValue);
    return this;
  }


  public AttributeBuilder setAttributeStructureType(String structureType) {
    this.structureType = structureType;
    return this;
  }

  public boolean isComplete() {
    return this.attributeName != null && this.attributeValues != null
        && this.structureType != null;
  }

  public Attribute<?> build(GenericAttributeFactory attributeFactory) {
    return attributeFactory.create(
        this.attributeName,
        this.structureType,
        this.attributeValues,
        new HashMap<>()
    );
  }
}
