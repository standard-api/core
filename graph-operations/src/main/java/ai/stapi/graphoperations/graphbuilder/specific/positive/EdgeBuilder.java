package ai.stapi.graphoperations.graphbuilder.specific.positive;

public interface EdgeBuilder extends GraphElementBuilder {

  EdgeDirection getEdgeDirection();

  EdgeBuilder setEdgeDirection(EdgeDirection edgeDirection);
}
