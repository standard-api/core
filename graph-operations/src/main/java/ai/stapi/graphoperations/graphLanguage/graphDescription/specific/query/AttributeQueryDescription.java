package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AbstractAttributeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AttributeDescriptionParameters;
import java.util.ArrayList;
import java.util.List;

public class AttributeQueryDescription extends AbstractAttributeDescription implements QueryGraphDescription {

  public static final String SERIALIZATION_TYPE = "AttributeQueryDescription";
  public static final String DESCRIBED_STRUCTURE_TYPE = "Irrelevant";

  private AttributeQueryDescription() {
  }

  public AttributeQueryDescription(
      AttributeDescriptionParameters parameters,
      GraphDescription... childDescriptions
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, parameters, childDescriptions);
  }

  public AttributeQueryDescription(
      AttributeDescriptionParameters parameters,
      List<GraphDescription> childDescriptions
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, parameters, childDescriptions);
  }

  public AttributeQueryDescription(
      String attributeName,
      List<GraphDescription> childDescriptions
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, attributeName, childDescriptions);
  }

  public AttributeQueryDescription(
      String attributeName
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, attributeName, new ArrayList<>());
  }
}
