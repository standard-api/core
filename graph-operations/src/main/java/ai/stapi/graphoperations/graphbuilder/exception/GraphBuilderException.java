package ai.stapi.graphoperations.graphbuilder.exception;

import ai.stapi.graphoperations.graphbuilder.specific.positive.NodeBuilder;
import ai.stapi.graphoperations.graphbuilder.specific.positive.GraphElementBuilder;

public class GraphBuilderException extends RuntimeException {

  private GraphBuilderException(String message) {
    super(message);
  }

  public static GraphBuilderException becauseAttributeValueTypeIsNotSupported(Object value) {
    return new GraphBuilderException(
        "Attribute value of type '" + value.toString() + "' is not supported.");
  }

  public static GraphBuilderException becauseAttributeBuilderIsNotComplete() {
    return new GraphBuilderException("Attribute builder is not complete.");
  }

  public static GraphBuilderException becauseNodeIsMissingIdOrType() {
    return new GraphBuilderException("Node is missing id or type.");
  }

  public static GraphBuilderException becauseEdgeIsNotCompleted() {
    return new GraphBuilderException(
        "Edge was not complete. It is missing either id, type or direction.");
  }

  public static GraphBuilderException becauseNodeFromOnEdgeIsNotComplete() {
    return new GraphBuilderException("Node FROM on edge is not complete.");
  }

  public static GraphBuilderException becauseNodeToOnEdgeIsNotComplete() {
    return new GraphBuilderException("Node TO on edge is not complete.");
  }

  public static GraphBuilderException becauseGraphCanNotHaveEdgeAsFirstElement() {
    return new GraphBuilderException("Graph can not have edge as first element.");
  }

  public static GraphBuilderException becauseGraphCanNotHaveEdgeAsLastElement() {
    return new GraphBuilderException("Graph can not have edge as last element.");
  }

  public static GraphBuilderException becauseOfUnexpectedEdgeNeighbor(
      GraphElementBuilder elementBuilder) {
    return new GraphBuilderException(
        "Because there was unexpected edge neighbor: '"
            + (elementBuilder == null ? "null'."
            : elementBuilder.getClass().getSimpleName() + "'."));
  }

  public static GraphBuilderException becauseThereCanNotBeTwoNodesAfterEachOther(
      NodeBuilder previousNode) {
    return new GraphBuilderException(
        "There can not be two nodes after each other. Previous node type: '"
            + previousNode.getType() + "'."
    );
  }

  public static GraphBuilderException becauseThereCanNotBeTwoEdgesAfterEachOther() {
    return new GraphBuilderException("There can not be two edges after each other.");
  }

  public static GraphBuilderException becauseEdgeHasToBePrecededByNode() {
    return new GraphBuilderException("Edge has to be preceded by node.");
  }

  public static GraphBuilderException becauseThereAreNoElementsAbleToContainAttribute() {
    return new GraphBuilderException("There are no elements able to contain attributes.");
  }

  public static GraphBuilderException becauseThereAreNoAttributesOnElement() {
    return new GraphBuilderException("There are no attributes on element.");
  }

  public static GraphBuilderException becauseThereAreNoElements() {
    return new GraphBuilderException("There are no elements inside builder.");
  }
}
