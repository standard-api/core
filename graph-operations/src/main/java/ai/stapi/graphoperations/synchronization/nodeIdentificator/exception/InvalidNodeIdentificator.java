package ai.stapi.graphoperations.synchronization.nodeIdentificator.exception;

import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.PositiveGraphDescription;
import ai.stapi.utils.Stringifier;
import java.util.List;
import java.util.stream.Collectors;

public class InvalidNodeIdentificator extends RuntimeException {

    private InvalidNodeIdentificator(String becauseMessage) {
        super("Node identificator was invalid, because " + becauseMessage);
    }
    
    public static InvalidNodeIdentificator becausePathToIdentifiactionValueDoesntStartWithEdgeOrAttributeDescription(
        List<PositiveGraphDescription> invalidPaths
    ) {
        return new InvalidNodeIdentificator(
            String.format(
                "path to identification value does not start with edge or attribute description. Invalid paths: %n%s",
                invalidPaths.stream().map(Stringifier::convertToString).collect(Collectors.joining("\n"))
            )
        );
    }

    public static InvalidNodeIdentificator becausePathToIdentifiactionValueIsNotSinglePath(
        List<PositiveGraphDescription> invalidPaths
    ) {
        return new InvalidNodeIdentificator(
            String.format(
                "path to identification value is not single path. Invalid paths: %n%s",
                invalidPaths.stream().map(Stringifier::convertToString).collect(Collectors.joining("\n"))
            )
        );
    }

    public static InvalidNodeIdentificator becausePathToIdentifiactionValueIsNotEndingWithUuidIdentityOrAttributeDescription(
        List<PositiveGraphDescription> invalidPaths
    ) {
        return new InvalidNodeIdentificator(
            String.format(
                "path to identification value is not ending with uuid identity or attribute description. " +
                    "Invalid paths: %n%s",
                invalidPaths.stream().map(Stringifier::convertToString).collect(Collectors.joining("\n"))
            )
        );
    }
}
