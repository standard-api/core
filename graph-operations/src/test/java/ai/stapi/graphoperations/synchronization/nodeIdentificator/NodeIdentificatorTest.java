package ai.stapi.graphoperations.synchronization.nodeIdentificator;

import ai.stapi.graphoperations.synchronization.nodeIdentificator.exception.InvalidNodeIdentificator;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.EdgeDescriptionParameters;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.NodeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.OutgoingEdgeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query.AttributeQueryDescription;
import ai.stapi.test.base.UnitTestCase;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;

class NodeIdentificatorTest extends UnitTestCase {

    @Test
    void itWillThrowExceptionWhenProvidedIdentifyingPathStartsWithNode() {
        Executable executable = () -> new NodeIdentificator(new NodeDescription("irrelevant"));
        this.thenExceptionMessageApproved(InvalidNodeIdentificator.class, executable);
    }

    @Test
    void itWillThrowExceptionWhenProvidedIdentifyingPathIsNotSingleBranch() {
        Executable executable = () -> new NodeIdentificator(
            new OutgoingEdgeDescription(
                new EdgeDescriptionParameters("irrelevantEdge"),
                new AttributeQueryDescription("oneAttribute"),
                new AttributeQueryDescription("secondAttribute")
            )
        );
        this.thenExceptionMessageApproved(InvalidNodeIdentificator.class, executable);
    }

    @Test
    void itWillThrowExceptionWhenProvidedIdentifyingPathIsNotEndingWithValueDescription() {
        Executable executable = () -> new NodeIdentificator(
            new OutgoingEdgeDescription(
                new EdgeDescriptionParameters("irrelevantEdge")
            )
        );
        this.thenExceptionMessageApproved(InvalidNodeIdentificator.class, executable);
    }
}