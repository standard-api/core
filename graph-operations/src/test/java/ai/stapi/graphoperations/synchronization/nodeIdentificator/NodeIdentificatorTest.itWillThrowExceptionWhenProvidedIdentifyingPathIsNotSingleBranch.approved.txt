Node identificator was invalid, because path to identification value is not single path. Invalid paths: 
{
  "serializationType" : "OutgoingEdgeDescription",
  "childGraphDescriptions" : [ {
    "serializationType" : "AttributeQueryDescription",
    "childGraphDescriptions" : [ ],
    "parameters" : {
      "attributeName" : "oneAttribute"
    },
    "graphTraversingType" : "attribute",
    "declarationType" : "PositiveGraphDescription",
    "describedAttributeStructureType" : "Irrelevant"
  }, {
    "serializationType" : "AttributeQueryDescription",
    "childGraphDescriptions" : [ ],
    "parameters" : {
      "attributeName" : "secondAttribute"
    },
    "graphTraversingType" : "attribute",
    "declarationType" : "PositiveGraphDescription",
    "describedAttributeStructureType" : "Irrelevant"
  } ],
  "parameters" : {
    "edgeType" : "irrelevantEdge"
  },
  "outgoing" : true,
  "ingoing" : false,
  "declarationType" : "PositiveGraphDescription",
  "graphTraversingType" : "edge"
}